{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "922d2c1b-de98-44d8-9d35-a42480dea520",
   "metadata": {},
   "source": [
    "# 1. Approach to Optimisation\n",
    "- Focus on the bottlenecks\n",
    "- Consider Data/ Task parallelism\n",
    "- Use optimised compiled tools\n",
    "- Timing bash\n",
    "    - { time [application] ; } 2> time.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdad9847-0805-46d7-ac10-0b41c61125db",
   "metadata": {},
   "source": [
    "# 2. Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64d4ef6a-2935-4537-bcfc-d4b9a368431f",
   "metadata": {},
   "source": [
    "## 2.1 QC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07234cb5-103c-448e-bbb7-bee60196a1d0",
   "metadata": {},
   "source": [
    "## 2.2 Trimming\n",
    "* Common tools are fastp, Trimmomatic, Cutadapt, and Trim_Galore [1]\n",
    "* But most common are fastp and Trim_galore as fastp is fast and simple while trim_galore also generate QC report with trimming and filtering (integrate cutadpt and fastqc into a single process)\n",
    "* fastp is faster than trim_galore"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6463f1d9-9c48-44fd-8f9c-37bf90b60f39",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 2.3 Mapping and Quantification\n",
    "* By Align to genome + Quantify post alignment OR Quasi-mapping Approach\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "092bbcce-558e-4374-af86-3d2a35cbfac0",
   "metadata": {},
   "source": [
    "### 2.3.1 Alignment\n",
    "* Bowtie, Bowtie2, TopHat, TopHat2, HISAT, HISAT2, STAR are commonly used tools for alignment\n",
    "* HISAT2 demonstrated the quickest processing time\n",
    "* NOTE: Different software give different alignment rate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb76991e-2069-416f-8fb0-ebbcbe52c44d",
   "metadata": {},
   "source": [
    "### 2.3.2 Quantification\n",
    "* Alignment-based transcript quantification tools: featureCounts, HTseq, RSEM\n",
    "    * featureCounts is fast\n",
    "    * RSEM can simultaneously perform both alignment and quantification tasks\n",
    "* alignment-free transcript quantification tools, including Salmon, Kallisto\n",
    "* NOTE: Previous studies have demonstrated that alignment-based quantification tools yield higher accuracy in results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c41f7d17-390f-4339-9fc0-e6970a2efd25",
   "metadata": {},
   "source": [
    "## 2.4 Compiling abundance into count matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f2ec3ef-467d-4620-ad73-528a15fd53c5",
   "metadata": {},
   "source": [
    "# 3. Original seq bash script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2484909-9538-4878-8fd2-e26b6f5eadba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#PBS -N RNAseq\n",
    "#PBS -j oe\n",
    "#PBS -l select=1:ncpus=8:mem=40gb\n",
    "#PBS -l walltime=12:00:00\n",
    "\n",
    "#################### DO NOT MODIFY ###################################\n",
    "source /etc/profile.d/rec_modules.sh #make module list available\n",
    "bash\n",
    ". ~/.bashrc\n",
    "\n",
    "conda activate my_env\n",
    "#################### DO NOT MODIFY ###################################\n",
    "\n",
    "# define source dir\n",
    "SOURCE=\"/hpctmp/e1324266/5004/raw/download/fastq_files\"\n",
    "# define ref dir\n",
    "REF_PATH=\"/hpctmp/e1324266/5004/ref/paper\"\n",
    "# define output dir\n",
    "OUTPUT=\"/hpctmp/e1324266/5004/seq_bash\"\n",
    "# define results\n",
    "RESULTS=\"/hpctmp/e1324266/5004/seq_bash/results\"\n",
    "\n",
    "# define ref files\n",
    "genome=\"${REF_PATH}/GRCh37.p12.genome.fa\"\n",
    "gtf=\"${REF_PATH}/gencode.v18.annotation.gtf\"\n",
    "transcript=\"${REF_PATH}/gencode.v18.pc_transcripts.fa\"\n",
    "star_index=\"${REF_PATH}/star_index\"\n",
    "\n",
    "# deinfe dir\n",
    "fastqc_dir=\"${RESULTS}/results/01-fastqc\"\n",
    "trim_dir=\"${RESULTS}/results/03-trim\"\n",
    "align_dir=\"${RESULTS}/results/04-align\"\n",
    "quant_dir=\"${RESULTS}/results/05-quant-align\"\n",
    "\n",
    "# Make directory\n",
    "mkdir -p \"${RESULTS}\"\n",
    "mkdir -p \"${start_index}\"\n",
    "mkdir -p \"${fastqc_dir}\"\n",
    "mkdir -p \"${trim_dir}\"\n",
    "mkdir -p \"${align_dir}\"\n",
    "mkdir -p \"${quant_dir}\"\n",
    "\n",
    "source /app1/ebenv\n",
    "module load STAR/2.7.10b-GCC-11.3.0\n",
    "echo \"Starting creating STAR index\"\n",
    "## Creating index\n",
    "{ time STAR --runMode genomeGenerate \\\n",
    "    --runThreadN 8 \\\n",
    "    --genomeDir \"$star_index\" \\\n",
    "    --genomeFastaFiles \"$genome\" \\\n",
    "    --sjdbGTFfile \"$gtf\" \\\n",
    "    --sjdbOverhang 100; } 2> ${OUTPUT}/profiling_star_index.txt\n",
    "module unload STAR/2.7.10b-GCC-11.3.0\n",
    "\n",
    "time (\n",
    "    for fq in ${SOURCE}/*_1.fastq.gz\n",
    "    do\n",
    "            # grab base of filename for naming outputs\n",
    "            sample_id=$(basename $fq _1.fastq.gz)\n",
    "            echo \"Sample name is $sample_id\"\n",
    "\n",
    "    ##Paired fastq files\n",
    "            fq1=${SOURCE}/${sample_id}_1.fastq.gz\n",
    "            fq2=${SOURCE}/${sample_id}_2.fastq.gz\n",
    "\n",
    "    # set up output filenames and locations\n",
    "            #the trimming fastq format might be slight differnt due to different versions of the software\n",
    "            trim_out_fq1=${trim_dir}/${sample_id}_val_1.fq.gz\n",
    "            trim_out_fq2=${trim_dir}/${sample_id}_val_2.fq.gz\n",
    "\n",
    "            # BAM file\n",
    "            bam=${align_dir}/${sample_id}_Aligned.toTranscriptome.out.bam\n",
    "\n",
    "            #starting\n",
    "            echo \"Processing sample $sample_id\"\n",
    "    # Run FastQC\n",
    "            source /app1/ebenv\n",
    "            module load FastQC/0.12.1-Java-11\n",
    "            echo \"Starting QC for $sample_id\"\n",
    "           { time fastqc -t 8 -o \"$fastqc_dir\" -f fastq -q \"$fq1\" \"$fq2\"; } 2>> ${RESULTS}/profiling_fastqc.txt\n",
    "    # Run Trimming\n",
    "            source /app1/ebenv\n",
    "            module unload FastQC/0.12.1-Java-11 \n",
    "            module load Trim_Galore/0.6.10-GCCcore-11.3.0\n",
    "            echo \"Starting Trimming for $sample_id\"\n",
    "           { time trim_galore --cores 8 --gzip --paired \"$fq1\" \"$fq2\" -o \"$trim_dir\" --basename $sample_id; } 2>> ${RESULTS}/profiling_trim_run.txt\n",
    "\n",
    "    # Run STAR\n",
    "            source /app1/ebenv\n",
    "            module unload Trim_Galore/0.6.10-GCCcore-11.3.0\n",
    "            module load STAR/2.7.10b-GCC-11.3.0\n",
    "            echo \"Starting run star read mapping for $sample_id\"\n",
    "           # Run alignment with STAR\n",
    "           { time STAR --runThreadN 8 \\\n",
    "                --genomeDir \"$star_index\" \\\n",
    "                --readFilesIn \"$trim_out_fq1\" \"$trim_out_fq2\" \\\n",
    "                --readFilesCommand zcat \\\n",
    "                --outFileNamePrefix \"${align_dir}/${sample_id}_\" \\\n",
    "                --outFilterType BySJout \\\n",
    "                --outSAMtype BAM SortedByCoordinate \\\n",
    "                --outSAMattributes Standard \\\n",
    "                --sjdbGTFfile \"$gtf\" \\\n",
    "                --quantMode TranscriptomeSAM; } 2>> ${RESULTS}/profiling_star_align_run.txt\n",
    "\n",
    "    # RUN SALMON QUANT     \n",
    "            source /app1/ebenv\n",
    "            module unload STAR/2.7.10b-GCC-11.3.0\n",
    "            module load Salmon/1.9.0-GCC-11.3.0\n",
    "            echo \"Starting salmon quant for $sample_id\"\n",
    "            { time salmon quant --threads 8 \\\n",
    "                --targets $transcript \\\n",
    "                --libType A \\\n",
    "                --gencode \\\n",
    "                --output ${quant_dir}/${sample_id} \\\n",
    "                --alignments $bam \\\n",
    "                --seqBias \\\n",
    "                --useVBOpt; } 2>> ${RESULTS}/profiling_salmon_quant_run.txt   \n",
    "\n",
    "    done\n",
    ") 2> ${RESULTS}/profiling_loop.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dae527b7-93c5-4dc5-9d67-64425cabc8c5",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 4. Optim program seq bash script\n",
    "Package optimisation\n",
    "- fastqc\n",
    "- fastp\n",
    "- histat2\n",
    "- featurecounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae6f92cf-d877-4f23-8669-863075a39a36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# HISAT2 ref builder\n",
    "#!/bin/bash\n",
    "#PBS -N hisat2\n",
    "#PBS -l walltime=04:00:00\n",
    "#PBS -l select=1:ncpus=4:mem=16gb\n",
    "\n",
    "# Load required modules\n",
    "source /app1/ebenv\n",
    "module load HISAT2/2.2.1-gompi-2023a\n",
    "\n",
    "# Change to the directory where the job was submitted\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "REF_PATH=\"/hpctmp/e1324266/5004/ref/paper\"\n",
    "OUTPUT=\"/hpctmp/e1324266/5004/ref/hisat2_index/genome\"\n",
    "mkdir -p ${OUTPUT}\n",
    "genome=\"${REF_PATH}/GRCh37.p12.genome.fa\"\n",
    "hisat2-build ${genome} ${OUTPUT}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7300506e-a0d8-4348-a7b8-80929fee0afd",
   "metadata": {},
   "source": [
    "## 4.1 unpararellised workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d1485b7-b137-4ca9-99eb-901f2d8cc4a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#PBS -N rna_optim_1\n",
    "#PBS -j oe\n",
    "#PBS -l select=1:ncpus=12:mem=24gb\n",
    "#PBS -l walltime=12:00:00\n",
    "\n",
    "# Change to the directory where the job was submitted\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "# define source dir\n",
    "SOURCE=\"/hpctmp/e1324266/5004/raw/download/fastq_files\"\n",
    "# define ref dir\n",
    "REF_PATH=\"/hpctmp/e1324266/5004/ref/paper\"\n",
    "# define results\n",
    "RESULTS=\"/hpctmp/e1324266/5004/seq_optim_bash\"\n",
    "\n",
    "# define ref files\n",
    "genome=\"${REF_PATH}/GRCh37.p12.genome.fa\"\n",
    "gtf=\"${REF_PATH}/gencode.v18.annotation.gtf\"\n",
    "transcript=\"${REF_PATH}/gencode.v18.pc_transcripts.fa\"\n",
    "\n",
    "# define hisat2 index files\n",
    "hisat2_index=\"/hpctmp/e1324266/5004/ref/hisat2_index/genome\"\n",
    "\n",
    "# deinfe dir\n",
    "fastqc_dir=\"${RESULTS}/results/01-fastqc\"\n",
    "trim_dir=\"${RESULTS}/results/03-trim\"\n",
    "align_dir=\"${RESULTS}/results/04-align\"\n",
    "quant_dir=\"${RESULTS}/results/05-quant-align\"\n",
    "\n",
    "# Make directory\n",
    "mkdir -p \"${fastqc_dir}\"\n",
    "mkdir -p \"${trim_dir}\"\n",
    "mkdir -p \"${align_dir}\"\n",
    "mkdir -p \"${quant_dir}\"\n",
    "\n",
    "time (\n",
    "    for fq in ${SOURCE}/*_1.fastq.gz\n",
    "    do\n",
    "    # grab base of filename for naming outputs\n",
    "        sample_id=$(basename $fq _1.fastq.gz)\n",
    "        echo \"Sample name is $sample_id\"\n",
    "\n",
    "    ##Paired fastq files\n",
    "        fq1=${SOURCE}/${sample_id}_1.fastq.gz\n",
    "        fq2=${SOURCE}/${sample_id}_2.fastq.gz\n",
    "\n",
    "    # set up output filenames and locations\n",
    "    #the trimming fastq format might be slight differnt due to different versions of the software\n",
    "        trim_out_fq1=${trim_dir}/${sample_id}_trimmed_1.fq.gz\n",
    "        trim_out_fq2=${trim_dir}/${sample_id}_trimmed_2.fq.gz\n",
    "\n",
    "    # SAM file\n",
    "        sam=${align_dir}/${sample_id}.sam\n",
    "\n",
    "    #starting\n",
    "        echo \"Processing sample $sample_id\"\n",
    "        \n",
    "  # Run FastQC\n",
    "        source /app1/ebenv\n",
    "        module load FastQC/0.12.1-Java-11\n",
    "        echo \"Starting QC for $sample_id\"\n",
    "        { time fastqc -t 4 -o \"$fastqc_dir\" -f fastq -q \"$fq1\" \"$fq2\"; } 2>> ${RESULTS}/optim1_fastqc.txt\n",
    "    # Run Trimming\n",
    "        source /app1/ebenv\n",
    "        module unload FastQC/0.12.1-Java-11 \n",
    "        module load fastp/0.23.2-GCC-11.3.0\n",
    "        echo \"Starting Trimming for $sample_id\"\n",
    "        { time fastp --thread 6 \\\n",
    "                    --in1 \"$fq1\" \\\n",
    "                    --in2 \"$fq2\" \\\n",
    "                    --out1 ${trim_dir}/${sample_id}_trimmed_1.fq.gz \\\n",
    "                    --out2 ${trim_dir}/${sample_id}_trimmed_2.fq.gz \\\n",
    "                    --detect_adapter_for_pe \\\n",
    "                    --qualified_quality_phred 20; } 2>> ${RESULTS}/optim1_trim_run.txt\n",
    "        # Run HISAT2\n",
    "        source /app1/ebenv\n",
    "        module unload fastp/0.23.2-GCC-11.3.0\n",
    "        module load HISAT2/2.2.1-gompi-2023a\n",
    "        { time hisat2 -p 12 -x ${hisat2_index} -1 \"${trim_out_fq1}\" -2 \"${trim_out_fq2}\" -S \"${align_dir}/${sample_id}.sam\"; } 2>> ${RESULTS}/optim1_hisat2_align_run.txt\n",
    "\n",
    "        # RUN featureCounts QUANT     \n",
    "        source /app1/ebenv\n",
    "        module unload HISAT2/2.2.1-gompi-2023a\n",
    "        module load Subread/2.0.4-GCC-11.3.0\n",
    "        echo \"Starting featurecounts quant for $sample_id\"\n",
    "        { time featureCounts -p --countReadPairs -a ${gtf} -T 12 -t exon -g gene_id -M -o ${quant_dir}/${sample_id}_counts.txt ${sam}; } 2>> ${RESULTS}/optim1_subread_quant_run.txt   \n",
    "\n",
    "    done\n",
    ") 2> ${RESULTS}/optim1_loop.txt\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ad245a6-c90c-4f08-812f-88223afcef5d",
   "metadata": {},
   "source": [
    "## 4.2 parallelised\n",
    "* task parallelised --> fastqc in backgorund with \"&\" \"wait\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ab1e8cf-732f-401d-bce3-337889746502",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "# define source dir\n",
    "SOURCE=\"/hpctmp/e1324266/5004/raw/download/fastq_files\"\n",
    "# define ref dir\n",
    "REF_PATH=\"/hpctmp/e1324266/5004/ref/paper\"\n",
    "# define results\n",
    "RESULTS=\"/hpctmp/e1324266/5004/par_optim_bash\"\n",
    "\n",
    "# define ref files\n",
    "genome=\"${REF_PATH}/GRCh37.p12.genome.fa\"\n",
    "gtf=\"${REF_PATH}/gencode.v18.annotation.gtf\"\n",
    "transcript=\"${REF_PATH}/gencode.v18.pc_transcripts.fa\"\n",
    "\n",
    "# define hisat2 index files\n",
    "hisat2_index=\"/hpctmp/e1324266/5004/ref/hisat2_index/genome\"\n",
    "\n",
    "# deinfe dir\n",
    "fastqc_dir=\"${RESULTS}/results/01-fastqc\"\n",
    "trim_dir=\"${RESULTS}/results/03-trim\"\n",
    "align_dir=\"${RESULTS}/results/04-align\"\n",
    "quant_dir=\"${RESULTS}/results/05-quant-align\"\n",
    "\n",
    "# Make directory\n",
    "mkdir -p \"${fastqc_dir}\"\n",
    "mkdir -p \"${trim_dir}\"\n",
    "mkdir -p \"${align_dir}\"\n",
    "mkdir -p \"${quant_dir}\"\n",
    "\n",
    "\n",
    "for fq in ${SOURCE}/*_1.fastq.gz\n",
    "do\n",
    "# grab base of filename for naming outputs\n",
    "sample_id=$(basename $fq _1.fastq.gz)\n",
    "\n",
    "##Paired fastq files\n",
    "fq1=${SOURCE}/${sample_id}_1.fastq.gz\n",
    "fq2=${SOURCE}/${sample_id}_2.fastq.gz\n",
    "\n",
    "# set up output filenames and locations\n",
    "#the trimming fastq format might be slight differnt due to different versions of the software\n",
    "trim_out_fq1=${trim_dir}/${sample_id}_trimmed_1.fq.gz\n",
    "trim_out_fq2=${trim_dir}/${sample_id}_trimmed_2.fq.gz\n",
    "\n",
    "# SAM file\n",
    "sam=${align_dir}/${sample_id}.sam\n",
    "\n",
    "qsub << EOF\n",
    "\n",
    "#!/bin/bash\n",
    "#PBS -N rna_optim_2_${sample_id}\n",
    "#PBS -V\n",
    "#PBS -j oe\n",
    "#PBS -l select=1:ncpus=6:mem=12gb\n",
    "#PBS -l walltime=12:00:00\n",
    "\n",
    "#################### DO NOT MODIFY ###################################\n",
    "\n",
    "# Change to the directory where the job was submitted\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "#################### DO NOT MODIFY ###################################\n",
    "\n",
    "time (\n",
    "    echo \"Sample name is $sample_id\"\n",
    "\n",
    "    #starting\n",
    "    echo \"Processing sample $sample_id\"\n",
    "    # Run FastQC\n",
    "    source /app1/ebenv\n",
    "    module load FastQC/0.12.1-Java-11\n",
    "    echo \"Starting QC for $sample_id\"\n",
    "    { time fastqc -t 4 -o ${fastqc_dir} -f fastq -q ${fq1} ${fq2}; } 2>> ${RESULTS}/optim2_fastqc.txt &\n",
    "    # Run Trimming\n",
    "    source /app1/ebenv\n",
    "    module unload FastQC/0.12.1-Java-11 \n",
    "    module load fastp/0.23.2-GCC-11.3.0\n",
    "    echo \"Starting Trimming for $sample_id\"\n",
    "    { time fastp --thread 4 \\\n",
    "                --in1 ${fq1} \\\n",
    "                --in2 ${fq2} \\\n",
    "                --out1 ${trim_dir}/${sample_id}_trimmed_1.fq.gz \\\n",
    "                --out2 ${trim_dir}/${sample_id}_trimmed_2.fq.gz \\\n",
    "                --detect_adapter_for_pe \\\n",
    "                --qualified_quality_phred 20; } 2>> ${RESULTS}/optim2_trim_run.txt\n",
    "    \n",
    "    # Run HISAT2\n",
    "    source /app1/ebenv\n",
    "    module unload fastp/0.23.2-GCC-11.3.0\n",
    "    module load HISAT2/2.2.1-gompi-2023a\n",
    "    { time hisat2 -p 6 -x ${hisat2_index} \\\n",
    "           -1 ${trim_out_fq1} \\\n",
    "           -2 ${trim_out_fq2} \\\n",
    "           -S ${align_dir}/${sample}.sam; } 2>> ${RESULTS}/optim2_hisat2_align_run.txt\n",
    "\n",
    "    # RUN featureCounts QUANT     \n",
    "    source /app1/ebenv\n",
    "    module unload HISAT2/2.2.1-gompi-2023a\n",
    "    module load Subread/2.0.4-GCC-11.3.0\n",
    "    echo \"Starting featurecounts quant for $sample_id\"\n",
    "    { time featureCounts -p --countReadPairs \\\n",
    "                  -a ${gtf} \\\n",
    "                  -T 6 \\\n",
    "                  -t exon \\\n",
    "                  -g gene_id \\\n",
    "                  -M \\\n",
    "                  -o ${quant_dir}/${sample}_counts.txt \\\n",
    "                  ${sam}; } 2>> ${RESULTS}/optim2_subread_quant_run.txt   \n",
    "\n",
    "    wait\n",
    ") 2>> ${RESULTS}/${sample_id}_optim2_run.txt\n",
    "EOF\n",
    "\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30790d10-cced-4fcc-b9fc-50360861d9a8",
   "metadata": {},
   "source": [
    "# 5. Nextflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1560d4f-0ded-4257-a858-108d3150a956",
   "metadata": {},
   "outputs": [],
   "source": [
    "# main pbs script for qsub\n",
    "#!/bin/bash\n",
    "#PBS -N nextflow_pipeline\n",
    "#PBS -o nextflow_output.log\n",
    "#PBS -e nextflow_error.log\n",
    "#PBS -l walltime=24:00:00\n",
    "#PBS -l select=1:ncpus=12:mem=24gb\n",
    "\n",
    "# Load required modules\n",
    "source /app1/ebenv\n",
    "module load Nextflow\n",
    "\n",
    "# Change to the directory where the job was submitted\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "mkdir -p /hpctmp/e1324266/5004/nf/results\n",
    "\n",
    "# Run Nextflow\n",
    "{ time nextflow run main.nf -c nextflow.config -profile module; } 2> /hpctmp/e1324266/5004/nf/results/nf_total_time.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7911d0fc-df9c-4194-90eb-7b106b50b24f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nextflow.config file\n",
    "// CLI\n",
    "dag.enabled = true\n",
    "report.enabled = true\n",
    "timeline.enabled = true\n",
    "\n",
    "// Set up scheduler\n",
    "process {\n",
    "    process {\n",
    "    executor = 'pbspro'\n",
    "    clusterOptions = '-V'\n",
    "    }\n",
    "}\n",
    "\n",
    "// Set up resources\n",
    "process {\n",
    "\n",
    "    withLabel: low {\n",
    "                cpus = 4\n",
    "                memory = 8.GB\n",
    "    }\n",
    "\n",
    "    withLabel: high {\n",
    "                cpus = 6\n",
    "                memory = 12.GB\n",
    "    }\n",
    "}\n",
    "\n",
    "//configuration_profiles\n",
    "\n",
    "profiles {\n",
    "\n",
    "    module {\n",
    "                process {\n",
    "\n",
    "                        withName: FASTQC {\n",
    "                                beforeScript = '''\n",
    "                                source /app1/ebenv\n",
    "                                module load FastQC/0.12.1-Java-11\n",
    "                        '''\n",
    "                        }\n",
    "\n",
    "                        withName: TRIM {\n",
    "                                beforeScript = '''\n",
    "                                source /app1/ebenv\n",
    "                                module load fastp/0.23.2-GCC-11.3.0 \n",
    "                        '''\n",
    "                        }\n",
    "\n",
    "                        withName: HISAT2 {\n",
    "                                beforeScript = '''\n",
    "                                source /app1/ebenv\n",
    "                                module load HISAT2/2.2.1-gompi-2023a\n",
    "                        '''\n",
    "                        }\n",
    "\n",
    "                        withName: FEATURECOUNTS {\n",
    "                                beforeScript = '''\n",
    "                                source /app1/ebenv\n",
    "                                module load Subread/2.0.4-GCC-11.3.0\n",
    "                        '''\n",
    "                        }\n",
    "                }\n",
    "\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdbf0a38-17c2-4d73-a560-e0ffdd69fc57",
   "metadata": {},
   "outputs": [],
   "source": [
    "nextflow.enable.dsl=2\n",
    "\n",
    "params.input_dir = \"/hpctmp/e1324266/5004/raw/download/fastq_files\"\n",
    "params.output_dir = \"/hpctmp/e1324266/5004/nf/results\"\n",
    "params.fastqc_dir = \"${params.output_dir}/00-fastqc\"\n",
    "params.trim_dir = \"${params.output_dir}/01-trim\"\n",
    "params.align_dir = \"${params.output_dir}/02-align\"\n",
    "params.quant_dir = \"${params.output_dir}/03-quant\"\n",
    "\n",
    "params.ref_dir = \"/hpctmp/e1324266/5004/ref/paper\"\n",
    "params.annotation = \"${params.ref_dir}/gencode.v18.annotation.gtf\"\n",
    "params.genome = \"${params.ref_dir}/GRCh37.p12.genome.fa\"\n",
    "params.transcript = \"${params.ref_dir}/gencode.v18.pc_transcripts.fa\"\n",
    "params.index = \"/hpctmp/e1324266/5004/ref/hisat2_index/genome\"\n",
    "\n",
    "\n",
    "// Define processes\n",
    "process FASTQC {\n",
    "    publishDir params.fastqc_dir, mode:'copy'\n",
    "\n",
    "    label 'low'\n",
    "    \n",
    "    input:\n",
    "    tuple val(sample_id), path(read1), path(read2)\n",
    "\n",
    "    output:\n",
    "    path(\"*\")\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    fastqc --threads ${task.cpus} $read1 $read2\n",
    "    \"\"\"\n",
    "}\n",
    "\n",
    "process TRIM {\n",
    "    // Output to params.trim_dir directory\n",
    "    publishDir params.trim_dir, mode: 'copy'\n",
    "    \n",
    "    label 'high'\n",
    "    \n",
    "    input:\n",
    "    tuple val(sample), path(read1), path(read2)\n",
    "\n",
    "    output:\n",
    "    tuple val(sample), path(\"${sample}_trimmed_1.fq.gz\"), path(\"${sample}_trimmed_2.fq.gz\")\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    fastp --thread ${task.cpus} --in1 $read1 --in2 $read2 --out1 ${sample}_trimmed_1.fq.gz --out2 ${sample}_trimmed_2.fq.gz --detect_adapter_for_pe --qualified_quality_phred 20\n",
    "    \"\"\"\n",
    "}\n",
    "\n",
    "process HISAT2 {\n",
    "    publishDir params.align_dir, mode: 'copy'\n",
    "    \n",
    "    label 'high'\n",
    "    \n",
    "    input:\n",
    "    tuple val(sample), path(read1), path(read2)\n",
    "    \n",
    "    output:\n",
    "    tuple val(sample), path(\"${sample}.sam\")\n",
    "    \n",
    "    script:\n",
    "    \"\"\"\n",
    "    hisat2 -p ${task.cpus} -x ${params.index} \\\n",
    "           -1 ${read1} \\\n",
    "           -2 ${read2} \\\n",
    "           -S ${sample}.sam\n",
    "    \"\"\"\n",
    "}\n",
    "\n",
    "process FEATURECOUNTS {\n",
    "    // Output to params.quant_dir\n",
    "    publishDir params.quant_dir, mode: 'copy'\n",
    "    \n",
    "    label 'high'\n",
    "    \n",
    "    input:\n",
    "    tuple val(sample), path(file)\n",
    "\n",
    "    output:\n",
    "    path \"*\"\n",
    "\n",
    "    script:\n",
    "    \"\"\"\n",
    "    featureCounts -p --countReadPairs \\\n",
    "                  -a ${params.annotation} \\\n",
    "                  -T ${task.cpus} \\\n",
    "                  -t exon \\\n",
    "                  -g gene_id \\\n",
    "                  -M \\\n",
    "                  -o ${sample}_counts.txt \\\n",
    "                  ${file}\n",
    "    \"\"\"\n",
    "}\n",
    "\n",
    "// Define the workflow\n",
    "workflow {\n",
    "ch = channel.fromFilePairs(\"${params.input_dir}/*_{1,2}.fastq.gz\",flat: true)\n",
    "FASTQC(ch)\n",
    "TRIM(ch)\n",
    "HISAT2(TRIM.out)\n",
    "FEATURECOUNTS(HISAT2.out)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb6d9ea-144d-4d09-986c-e64c87656a36",
   "metadata": {},
   "source": [
    "# 6. References\n",
    "\n",
    "* https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-024-10414-y\n",
    "* https://www.frontiersin.org/journals/plant-science/articles/10.3389/fpls.2021.657240/full\n",
    "* https://academic.oup.com/nargab/article/6/2/lqae040/7659592"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
